plugins/ccgx/fu-ccgx-dmc-firmware.c

```diff
@@ -0,0 +1,332 @@
+				gsize* segment_info_offset,
+				GError **error)
+{
+	FuCcgxDmcFirmware *self = FU_CCGX_DMC_FIRMWARE (firmware);
+	gsize fw_buffer_offset = 0;
+	gsize row_data_offset = 0;
+	gsize digestlen = DMC_HASH_SIZE;
+	guint8 hash[DMC_HASH_SIZE];
+	g_autoptr(GChecksum) csum = g_checksum_new (G_CHECKSUM_SHA256);
+
+	/* set row data offset in current image */
+	row_data_offset = self->row_data_offset_start + img_rcd->info_header.img_offset;
+	/* create segment record array in image record */
+	img_rcd->seg_records = g_ptr_array_new_with_free_func ((GFreeFunc) fu_ccgx_dmc_firmware_segment_record_free);
+	/* parse segment in image */
+	for (guint32 seg_num = 0; seg_num < img_rcd->info_header.num_img_segments; seg_num++) {
+		guint16 actual_row_size = 0;
+		g_autoptr(FuCcgxDmcFirmwareSegmentRecord) seg_rcd = NULL;
+		g_autofree guint8 *row_buff = NULL;
+		/* alloc segment record */
+		seg_rcd = g_new0 (FuCcgxDmcFirmwareSegmentRecord, 1);
+		/* set segment info offset */
+		fw_buffer_offset = *segment_info_offset;
+		/* read segment info  */
+		if (!fu_memcpy_safe ((guint8 *)&seg_rcd->info_header, sizeof(seg_rcd->info_header), 0x0, /* dst */
+			     fw_buff, fw_buff_size, fw_buffer_offset, sizeof(seg_rcd->info_header), error)) /* src */
+			return FALSE;
+		/* calculate actual row size */
+		actual_row_size = img_rcd->info_header.row_size * 64;
+		/* alloc row buffer */
+		row_buff = g_malloc0 (actual_row_size);
+		/* create data record array in segment record */
+		seg_rcd->data_records = g_ptr_array_new_with_free_func ((GFreeFunc) fu_ccgx_dmc_firmware_data_record_free);
+		/* read row data in segment */
+		for (int row = 0; row < seg_rcd->info_header.num_rows; row++) {
+			g_autoptr(FuCcgxDmcFirmwareDataRecord) data_rcd = NULL;
+			/* alloc data record */
+			data_rcd = g_new0 (FuCcgxDmcFirmwareDataRecord, 1);
+			/* set row data offset */
+			fw_buffer_offset = row_data_offset;
+			/* read row data */
+			if (!fu_memcpy_safe (row_buff, actual_row_size, 0x0, /* dst */
+			     fw_buff, fw_buff_size, fw_buffer_offset, actual_row_size, error)) /* src */
+				return FALSE;
+			/* update hash */
+			g_checksum_update (csum, (guchar *) row_buff, actual_row_size);
+			/* add row data to data record */
+			data_rcd->data = g_bytes_new (row_buff, actual_row_size);
+			/* add data record to data record array */
+			g_ptr_array_add (seg_rcd->data_records, g_steal_pointer (&data_rcd));
+			/* increment row data offset */
+			row_data_offset += actual_row_size;
+		}
+		/* add segment record to segment array */
+		g_ptr_array_add (img_rcd->seg_records, g_steal_pointer (&seg_rcd));
+		/* increment segment info offset */
+		*segment_info_offset += sizeof(FwctSegmentationInfo);
+	}
+
+	/* check checksum */
+	g_checksum_get_digest (csum, hash, &digestlen);
+	if (memcmp (hash, img_rcd->info_header.img_digest, DMC_HASH_SIZE) != 0) {
+		g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid hash");
```

Comment:
maybe add what was expected and obtained?

Reply:
HASH size is 32 byte. How to print 32 byte value?

Reply:
So maybe not super useful. We can add it later if we hit this.